swupd-extract/state.go

```diff
@@ -0,0 +1,367 @@
+
+import (
+	"archive/tar"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/clearlinux/mixer-tools/swupd"
+)
+
+type clientState struct {
+	dir string
+
+	noCache     bool
+	baseContent string
+	isRemote    bool
+}
+
+func newClientState(stateDir, baseContent string) (*clientState, error) {
+	contentFilename := filepath.Join(stateDir, "content")
+	stateContent, err := ioutil.ReadFile(contentFilename)
+	if err == nil {
+		if string(stateContent) != baseContent {
+			fmt.Printf("- Resetting state in %s previously used for %q\n", stateDir, stateContent)
+			// Delete files individually in case stateDir is managed by the user.
+			var fis []os.FileInfo
+			fis, err = ioutil.ReadDir(stateDir)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't reset state directory: %s", err)
+			}
+			for _, fi := range fis {
+				err = os.RemoveAll(filepath.Join(stateDir, fi.Name()))
+				if err != nil {
+					return nil, fmt.Errorf("couldn't reset state directory: %s", err)
+				}
+			}
+		}
+	}
+	err = os.MkdirAll(filepath.Join(stateDir, "staged/temp"), 0755)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create state directory: %s", err)
+	}
+	err = ioutil.WriteFile(contentFilename, []byte(baseContent), 0644)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't write %s: %s", contentFilename, err)
+	}
+
+	var isRemote bool
+	if strings.HasPrefix(baseContent, "https://") || strings.HasPrefix(baseContent, "http://") {
+		isRemote = true
+	}
+
+	cs := &clientState{
+		dir:         stateDir,
+		baseContent: baseContent,
+		isRemote:    isRemote,
+	}
+
+	return cs, nil
+}
+
+func (cs *clientState) OpenFile(elem ...string) (io.ReadCloser, error) {
+	joined := filepath.Join(elem...)
+	if !cs.isRemote {
+		return os.Open(filepath.Join(cs.baseContent, joined))
+	}
+
+	u := cs.baseContent + "/" + joined
+	fmt.Printf("- downloading %s\n", u)
+	res, err := http.Get(u)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != http.StatusOK {
+		_ = res.Body.Close()
+		return nil, fmt.Errorf("couldn't download %q: got response with code: %d %s", u, res.StatusCode, http.StatusText(res.StatusCode))
+	}
+	return res.Body, nil
+}
+
+func (cs *clientState) GetFile(elem ...string) (string, error) {
+	joined := filepath.Join(elem...)
+	if !cs.isRemote {
+		return filepath.Join(cs.baseContent, joined), nil
+	}
+	localFile := filepath.Join(cs.dir, joined)
+	if _, err := os.Stat(localFile); err == nil {
+		if !cs.noCache {
+			return localFile, nil
+		}
+		err = os.RemoveAll(localFile)
+		if err != nil {
+			return "", fmt.Errorf("couldn't remove %s to redownload: %s", localFile, err)
+		}
+	}
+	err := os.MkdirAll(filepath.Dir(localFile), 0755)
+	if err != nil {
+		return "", err
+	}
+	err = download(cs.baseContent+"/"+joined, localFile)
+	if err != nil {
+		return "", err
+	}
+	return localFile, nil
+}
+
+func (cs *clientState) Path(elem ...string) string {
+	return filepath.Join(cs.dir, filepath.Join(elem...))
```

Comment:
join of join?

Reply:
I can't do Join(cs.dir, elem...).

Reply:
I've considered also `filepath.Join(append([]string{cs.dir}, elem...)...)` but looks uglier. Maybe I'm missing a simpler solution here?

Reply:
I think this is fine... you can split into 
`elems := filepath.Join(elem...)`
``return filepath.Join(cd.dir, elems)` and waste another line of space, I can read it fine either way.